htlcswitch/link.go

```diff
@@ -1150,6 +1228,132 @@ func (l *channelLink) htlcManager() {
 				"PendingLocalUpdateCount")
 		}
 
+		// If we are attempting to shutdown the link to cooperatively
+		// close and the channel is clear of our pending updates and we
+		// haven't already sent a Shutdown, we'll send one now.
+		if (l.shutdownInit || l.shutdownReceived) && !l.shutdownSent &&
+			l.channel.PendingLocalUpdateCount() == 0 {
+
+			// Create the Shutdown we'll send over. Send the local
+			// upfront shutdown script if it exists, and the
+			// provided delivery address otherwise. This is used in
+			// case we did not initiate the coop close.
+			upfrontAddr := l.channel.LocalUpfrontShutdownScript()
+			if len(upfrontAddr) != 0 {
+				l.cfg.DeliveryAddr = upfrontAddr
+			} else {
+				// Persist the delivery script for
+				// retransmission.
+				err := l.channel.State().PersistDeliveryScript(
```

Comment:
tbh i'm a bit concerned about accessing methods like `l.channel.State().Persis...`, feels like there's a hidden nil panic waiting for us😂

Reply:
accessing `State()` is used in this file a lot, it could be improved but every link is initiated with a *channeldb.OpenChannel pointer